src/messaging/ReliableMessageContext.cpp

```diff
@@ -0,0 +1,458 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the CHIP reliable message protocol.
+ *
+ */
+
+#include <inttypes.h>
+
+#include <messaging/ReliableMessageContext.h>
+
+#include <core/CHIPEncoding.h>
+#include <messaging/ErrorCategory.h>
+#include <messaging/Flags.h>
+#include <messaging/ReliableMessageManager.h>
+#include <protocols/CHIPProtocols.h>
+#include <protocols/common/CommonProtocol.h>
+#include <support/BitFlags.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+namespace messaging {
+
+void ReliableMessageContextDeletor::Release(ReliableMessageContext * obj)
+{
+    obj->mDelegate.GetManager().FreeContext(obj);
+}
+
+ReliableMessageContext::ReliableMessageContext(ReliableMessageDelegate & delegate) :
+    mConfig(gDefaultReliableMessageProtocolConfig), mNextAckTimeTick(0), mThrottleTimeoutTick(0), mPendingPeerAckId(0),
+    mDelegate(delegate)
+{}
+
+/**
+ *  Determine whether there is already an acknowledgment pending to be sent
+ *  to the peer on this exchange.
+ *
+ */
+bool ReliableMessageContext::IsAckPending() const
+{
+    return GetFlag(mFlags, kFlagAckPending);
```

Comment:
nit: I would note that we deprecated GetFlag/SetFlag/ClearFlag in favor of BitFlags (typesafe version).

This would imply a bit of a refactor: flags would become enum values, members change a bit.

Reply:
done